### ISSUE TITLE
pgr_trsp on windows returns no route

### ISSUE THREAD

**Issue by Tho77** [pgRouting/pgrouting#336]
I've been using pgr_dijkstra successfully for several months and am now trying to switch to pgr_trsp.
I've been unable to get any sort of route back from pgr_trsp, so I started fresh and with the sample data, and am still unable to get any route back.

Installed a fresh Windows PostgreSQL v9.4.1 x64 on my local dev box, and used the Application Stack Builder to install PostGIS v2.1.7 which includes pgRouting 2.0.
Created the sample data from http://docs.pgrouting.org/2.0/en/doc/src/developer/sampledata.html
Running the basic sample queries from http://docs.pgrouting.org/2.0/en/src/dijkstra/doc/index.html return valid routes.
Running the basic sample queries from http://docs.pgrouting.org/2.0/en/src/trsp/doc/index.html just returns an empty route.

    seq  node  edge  cost
      0    -1     0     0
I initially tried it on some of our other Windows servers with PostgreSQL v9.3/9.4 where pgr_dijkstra works, but got back the same empty route for pgr_trsp.

I don't easily have a way to test a linux installation, but if you think that may be the issue, I will try and put one together to test.

**Comment by woodbri** [comment]
Create a test database and install pgrouting.
then fetch this file http://imaptools.com:8080/dl/parallel.sql
Open that in pgadmin3 in a sql window and execute it
then run this command:
```
select * from pgr_trsp('select * from parallel', 1, 4, false, true);
```
I get the following response on Linux:

```
 seq | id1 | id2 | cost
-----+-----+-----+------
   0 |   1 |   1 |    1
   1 |   2 |   2 |    2
   2 |   3 |   5 |    1
   3 |   4 |  -1 |    0
(4 rows)
```

**Comment by Tho77** [comment]
Thanks for such a quick response. Unfortunately my issue persists.

    select * from parallel;
![image](https://cloud.githubusercontent.com/assets/731788/7597497/f3889692-f8c1-11e4-9e13-7931190def60.png)

    select * from pgr_trsp('select * from parallel', 1, 4, false, true);
![image](https://cloud.githubusercontent.com/assets/731788/7597516/0f2aaed0-f8c2-11e4-8144-3a9e7221d219.png)

    select * from pgr_dijkstra('select * from parallel', 1, 4, false, true);
![image](https://cloud.githubusercontent.com/assets/731788/7597545/384f284a-f8c2-11e4-8a45-b25710e0222d.png)

**Comment by robe2** [comment]
woodbri - confirmed I get the same thing on my windows box.  Which pgrouting are you using woodbri.  Wonder if its difference between released and development.

Output for 

```
pgr=# SELECT * from pgr_version();
 version |       tag       | build |  hash   | branch | boost
---------+-----------------+-------+---------+--------+--------
 2.0.0   | pgrouting-2.0.0 | 0     | f26831f | master | 1.53.0
(1 row)
```
and this is on version 
```
pgr=# SELECT version();
                           version
-------------------------------------------------------------
 PostgreSQL 9.4.1, compiled by Visual C++ build 1800, 64-bit
```

I'll install the development version of pgrouting to see if I see the issue there, unless you are testing with the released version and don't see it.

**Comment by robe2** [comment]
Put in another ticket.  Having issues compiling the development branch noted here:

https://github.com/pgRouting/pgrouting/issues/337

**Comment by woodbri** [comment]
@robe2 I'm using:
```
pgrouting=# select * from pgr_version();
 version |       tag       | build |  hash   | branch  | boost
---------+-----------------+-------+---------+---------+--------
 2.0.0   | pgrouting-2.0.0 | 170   | 5cdd106 | develop | 1.54.0
(1 row)
```

**Comment by Tho77** [comment]
Managed to get a CentOS VM going, installed fresh PostgreSQL 9.4.1 and

    gisdb=# SELECT * from pgr_version();
    version |       tag       | build |  hash   | branch | boost
    ---------+-----------------+-------+---------+--------+--------
    2.0.0   | pgrouting-2.0.0 | 0     | f26831f | master | 1.53.0

pgr_trsp works correctly.

**Comment by woodbri** [comment]
Yeah, I'm not sure what is going on with this, it is working on Ubuntu 14.04 also, so I suspect that it is a windows specific problem which is a pain to sort out because I don't have my windows development environment setup up anymore. I'll try to work with @robe2 to get it sorted out but it might take a while.

**Comment by robe2** [comment]
woodbri, Tho:  Well the good news is that the build generated from Vicky's ksp improve branch
https://github.com/cvvergara/pgrouting/tree/ksp-improve-boost
 does behave correctly under windows 64-bit 9.4.1
```
 version |       tag       | build |  hash   | branch  | boost
---------+-----------------+-------+---------+---------+--------
 2.0.0   | pgrouting-2.0.0 | 170   | 5cdd106 | develop | 1.53.0
(1 row)
```
```
SELECT * FROM pgr_dijkstra('select * from parallel', 1, 4, false, true);
```
get output:
```
 seq | id1 | id2 | cost
-----+-----+-----+------
   0 |   1 |   1 |    1
   1 |   2 |   2 |    2
   2 |   3 |   5 |    1
   3 |   4 |  -1 |    0
```

You can download the build here:

http://winnie.postgis.net/download/windows/pg94/buildbot/  (the one called pgrouting-pg94-binaries-ksp_improve_boostw64gcc48.zip )

I still have to check the 9.3 because that one is throwing some errors which I think according to Vicky are related to trsp.

**Comment by Tho77** [comment]
Thanks for the suggestion @robe2 , I installed ksp-improve-boost on 9.4.1, getting the same pgr_version() as you.
pgr_dijkstra works, pgr_trsp still gives an empty result though.

    SELECT * FROM pgr_trsp('select * from parallel', 1, 4, false, true);

**Comment by robe2** [comment]
Ah bah I was testing the wrong thing.  Okay still a problem.  That was too easy I guess..

**Comment by robe2** [comment]
Okay this is one of those weird ones where it works fine in my mingw64 install, but doesn't on the VC++ build I install it on.  I was meaning to also have pgrouting test against the VC++ builds.  Anyway that's probably why it passes the tests fine but fails in the general environment.

woodbri -- do you support DEBUG level notices so I can switch my postgres to DEBUG level to track what is up here?

**Comment by woodbri** [comment]
DEBUG you can uncomment this line:
https://github.com/pgRouting/pgrouting/blob/master/src/trsp/src/trsp.c#L16
and it will turn on NOTICE from the C code.

**Comment by woodbri** [comment]
@robe2 look here:
https://github.com/pgRouting/pgrouting/blob/master/src/ksp/src/ksp.c#L316
@cvvergara found this and maybe this is the problem that trsp does not have the MSC ifdef like the link above.

**Comment by woodbri** [comment]
@robe2, I made a change to trsp.c in develop branch to try fix this issue and pushed it.  Ineed to build on winnie and the the user to test it out and see if it fixes the problem. I can make a change to master if it does and bump the version.

**Comment by woodbri** [comment]
@robe2 We just merged all of @cvvergara changes for ksp and dijkstra into develop. So we should only look at development for now.

**Comment by woodbri** [comment]
@Tho77 if you can pull a build of develop from postgis.net and see if the changes I added to develop resolve your problem, that would help. Thanks.

**Comment by Tho77** [comment]
Installed the one stamped today at 2:46pm.
http://winnie.postgis.net/download/windows/pg94/buildbot/pgrouting-pg94-binaries-2.0w64gcc48.zip
![image](https://cloud.githubusercontent.com/assets/731788/7661206/95ec3ae8-fb1e-11e4-825a-90180d8cf91c.png)

    SELECT * FROM pgr_trsp('select * from parallel', 1, 4, false, true);
is still empty.

**Comment by woodbri** [comment]
OK, thank you for checking. I'll have to leave this to @robe2 to figure out as it is some kind of windows compatibility or build issue.

**Comment by robe2** [comment]
Okay will take a look later this weekend.

**Comment by robe2** [comment]
This looks like a character encoding issue.  I put the debug line in, and when I ran on my default VC++ instance (which happens to default to WIN1252 ) it gave the below error, however if I create a new database using template0 with UTF-8 and run in pgAdmin it gives the right anwer.  However since my psql is set to win1252 encoding, it gives the same error below even on my utf8 database unless I do
```
\encoding UTF8
SELECT * FROM pgr_trsp('select * from parallel', 1, 4, false, true);

Gives expected:
 seq | id1 | id2 | cost
-----+-----+-----+------
   0 |   1 |   1 |    1
   1 |   2 |   2 |    2
   2 |   3 |   5 |    1
   3 |   4 |  -1 |    0
```
NOTICE:  Calling compute_trsp
NOTICE:  start turn_restrict_shortest_path
NOTICE:  columns: id 1 source 2 target 3 cost 4
NOTICE:  columns: reverse_cost cost 5
NOTICE:  Min vertex id: 1 , Max vid: 6
NOTICE:  Total 8 edge tuples
NOTICE:  Fetching restriction tuples
NOTICE:  Sql for restrictions is null.
NOTICE:  Total 0 restriction tuples
NOTICE:  Calling trsp_node_wrapper
NOTICE:  Message received from inside:


ERROR:  character with byte sequence 0x8d in encoding "WIN1252" has no equivalent in encoding "UTF8"
********** Error **********

ERROR: character with byte sequence 0x8d in encoding "WIN1252" has no equivalent in encoding "UTF8"
SQL state: 22P05

```

I did run the regress tests against VC++ and they all passed the last two times I ran, however on the first run, I got an error in bdstar which I haven't been able to repeat and forgot to log.

**Comment by robe2** [comment]
oops sorry launched the wrong instance.  Stay tuned. :(

**Comment by robe2** [comment]
Well it seems to work when I compile on my machine and install.  I'll repull winnie's binaries to see whats up.  This time I triple checked I was running on the right instance with:

```
pgr=# SELECT version();
                           version
-------------------------------------------------------------
 PostgreSQL 9.4.1, compiled by Visual C++ build 1800, 64-bit
(1 row)


pgr=# SELECT postgis_full_version();
NOTICE:  Function postgis_topology_scripts_installed() not found. Is topology su
pport enabled and topology.sql installed?

                                            postgis_full_version


--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
-----------------------------
 POSTGIS="2.2.0dev r13489" GEOS="3.5.0dev-CAPI-1.9.0 r4054" PROJ="Rel. 4.8.0, 6
March 2012" GDAL="GDAL 1.11.1, released 2014/09/24" LIBXML="2.7.8" LIBJSON="0.12
" (core procs from "2.2.0dev r13445" need upgrade) RASTER (raster procs from "2.
2.0dev r13445" need upgrade)
(1 row)


pgr=# SELECT * FROM pgr_version();
 version |       tag       | build |  hash   | branch  | boost
---------+-----------------+-------+---------+---------+--------
 2.0.0   | pgrouting-2.0.0 | 170   | 5cdd106 | develop | 1.53.0
(1 row)


pgr=# SELECT * FROM pgr_trsp('select * from parallel', 1, 4, false, true);
NOTICE:  Calling compute_trsp
NOTICE:  start turn_restrict_shortest_path

NOTICE:  columns: id 1 source 2 target 3 cost 4
NOTICE:  columns: reverse_cost cost 5
NOTICE:  Min vertex id: 1 , Max vid: 6
NOTICE:  Total 8 edge tuples
NOTICE:  Fetching restriction tuples

NOTICE:  Sql for restrictions is null.
NOTICE:  Total 0 restriction tuples
NOTICE:  Calling trsp_node_wrapper

NOTICE:  Message received from inside:
NOTICE:  XdB
NOTICE:  ret = 0

NOTICE:  *path_count = 4

NOTICE:  In finish, trying to disconnect from spi 0
NOTICE:  Ret is 0
NOTICE:  Total cost is: 4.000000
NOTICE:  Going to free path
 seq | id1 | id2 | cost
-----+-----+-----+------
   0 |   1 |   1 |    1
   1 |   2 |   2 |    2
   2 |   3 |   5 |    1
   3 |   4 |  -1 |    0
(4 rows)

```

and I get above.

@woodbri @dkastl  would really help at this point if one of you could update the hash on develop so I can tell if I'm testing the same thing and winnie is not getting a stale binary from somewhere.

**Comment by woodbri** [comment]
Just bumped the VERSION hash

**Comment by robe2** [comment]
Thanks hash  changed when I pulled winnies's build but gives broken answer.  So going to compare systems next.

**Comment by robe2** [comment]
Okay I may be crazy, but it seems setting the debug on was what fixed it on my system.  If I take it out, it doesn't work anymore.  Perhaps a timing issue and the debug notice is slowing it down enough so it can finish.

**Comment by woodbri** [comment]
More likely turning on DEBUG is enabling more code to compile which is changing the order of things in memory, and somebody is overwriting memory somewhere. Could this be an issue like you can not pass relative pointers to a DLL? 

At some point I'll run valgrind on trsp and see if that shows any memory read/write access errors. But I have my plate full at the moment.

**Comment by woodbri** [comment]
Then again I thought this used to work just fine on windows, but maybe not.

**Comment by robe2** [comment]
For reference -- I'm almost ready to switch winnie to testing against the VC+ build so will be easier to track issue fixes for #339 . I hope to be done with that today or tomorrow.  The last piece is I gotta have PostGIS install on the EDB (right now though PostGIS tests the EDB install, its not actually installing it since it runs it directly from PostGIS regress (instead of using extension machinery for testing).  I need to do that anyway to accommodate testing extension install.

**Comment by msgoloborodov** [comment]
Have same results on Ubuntu 13.10 (POSTGIS="2.1.7 r13414" GEOS="3.4.2-CAPI-1.8.2 r3921" PROJ="Rel. 4.7.1, 23 September 2009" GDAL="GDAL 1.11.2):

seq | node | edge | cost 
-----+------+------+------
   0 |   -1 |    0 |    0
(1 row)

**Comment by woodbri** [comment]
Please post the results of  ``select * from pgr_version();``

**Comment by msgoloborodov** [comment]
(2.0.0,pgrouting-2.0.0,280,0acd2ea,develop,1.54.0)

**Comment by woodbri** [comment]
I just tried this again and it worked using this branches: 
(2.0.0,pgrouting-2.0.0,95,c77eecf,develop_ksp,1.53.0) pg 9.3
(2.0.0,pgrouting-2.0.0,170,5cdd106,develop,1.54.0) pg 9.4
(2.0.0,pgrouting-2.0.0,286,b1d7d3c,develop,1.54.0)  pg 9.4
on my Ubuntu 14.04.2 LTS

If you checkout the repository from github, then run script ``tools/doit`` this will build the source, and run the test suite.

**Comment by msgoloborodov** [comment]
It doesn't work on my Ubuntu 13.10 running on virtual machine under Windows Server Hyper-V.
```
dorgis=# select * from pgr_version();
 version |       tag       | build |  hash   |   branch    | boost  
---------+-----------------+-------+---------+-------------+--------
 2.0.0   | pgrouting-2.0.0 | 95    | c77eecf | develop_ksp | 1.54.0
(1 row)

dorgis=#  SELECT seq, id1 AS node, id2 AS edge, cost
           FROM pgr_trsp(
                   'SELECT id, source, target, cost FROM edge_table',
                   7, 12, false, false,
                   'SELECT to_cost, to_edge AS target_id,
              from_edge || coalesce('','' || via, '''') AS via_path
          FROM restrictions'
           );
 seq | node | edge | cost 
-----+------+------+------
   0 |   -1 |    0 |    0
(1 row)
```

**Comment by manikanta-kondeti** [comment]
Hi, 
Can you check the database once?  Because its working for me, I just checked it yesterday.
Show the output of this commands. 
\d
select * from edge_table;  

Follow this if it not working.

**Comment by msgoloborodov** [comment]
pgr_trsp work fine when i don't use restriction sql:
```
# SELECT seq, id1 AS node, id2 AS edge, cost
#            FROM pgr_trsp(
#                    'SELECT id, source, target, cost FROM edge_table',
#                    7, 12, false, false);
 seq | node | edge | cost 
-----+------+------+------
   0 |    7 |    8 |    1
   1 |    8 |   11 |    1
   2 |   11 |   13 |    1
   3 |   12 |   -1 |    0
(4 rows)
```

But with restriction it always returns one row.

**Comment by manikanta-kondeti** [comment]
So did you try adding the restrictions data? 
CREATE TABLE restrictions (
    rid serial,
    to_cost double precision,
    to_edge integer,
    from_edge integer,
    via text
);

INSERT INTO restrictions VALUES (1,100,7,4,null);
INSERT INTO restrictions VALUES (2,4,8,3,5);
INSERT INTO restrictions VALUES (3,100,9,16,null);

**Comment by msgoloborodov** [comment]
with trsp-improvements build:
```
#  SELECT seq, id1 AS node, id2 AS edge, cost
           FROM pgr_trsp(
                   'SELECT id, source, target, cost FROM edge_table',
                   7, 12, false, false,
                   'SELECT to_cost, to_edge AS target_id,
              from_edge || coalesce('','' || via, '''') AS via_path
          FROM restrictions'
           );
 seq | node | edge | cost 
-----+------+------+------
   0 |   -1 |    0 |    0
(1 row)

# select * from restrictions;
 rid | to_cost | to_edge | from_edge | via 
-----+---------+---------+-----------+-----
   1 |     100 |       7 |         4 | 
   2 |       4 |       8 |         3 | 5
   3 |     100 |       9 |        16 | 
(3 rows)
```

**Comment by cvvergara** [comment]
Tagging it as discussion, becuse it looks like the original issue creator was not using the latest develop

**Comment by robe2** [comment]
As you can see - I just committed a change to trsp in develop_2_1_0 branch that (only tested on my PostgreSQL 9.4 EDB 64-bit EDB install).  That makes the code work if client_min_messages allows NOTICE or log_min_messages logs NOTICE.

https://github.com/pgRouting/pgrouting/commit/3edbc88109588e30df965a60c3556c9a6ecb5ed6#diff-5eea6139b9a23db16dca75d53f46351fR510

I don't know why that makes it work.  I still feel like maybe its slowing something down and was tempted to replace that with a sleep call to see if it still works.

There is only one section in the code where I had to throw up a notice, so I presume its somewhere around there that is the issue.  FWIW - turns out this not  an issue on the windows VC++ 32-bit installs (using the PostGIS bundle, just the 64-bit EDB distributed when used with my mingw builds).  

I'm not sure if the issue exists for native pgRouting VC++ builds  for 64-bit -- @sanak  maybe you can check that out since I think you build with VC++

**Comment by sanak** [comment]
@robe2 Okay, I will try to check this issue later.

**Comment by sanak** [comment]
@robe2
I checked my pgrouting4w native msvc build, but the error didn't occur.
So, I guess that this is Windows mingw build specific issue.

By the way, my mingw64 build box(environment) is old (around 2012),
so if you tell me current master/develop mingw64 box location,
I will be able to try to check what is the cause in mingw64 later.

**Comment by robe2** [comment]
@sanak 

For the pgrouting4w native, were you testing 32-bit or 64-bit or both.  The issue only happens with 64-bit build.

What I can do is zip up my whole mingw environment, that way you don't have to rebuild everything.  I need to get rid of some junk first to make it light weight though since I think the one I have up is a bit dated too (need to update it anyway).

**Comment by robe2** [comment]
I updated the download I have so has latest cgal, postgis, postgres, and other dependencies -needed to build everything: http://www.bostongis.com/postgisstuff/ming64gcc48.7z

You should be able to just extract and read the read me to get going.

**Comment by sanak** [comment]
@robe2 
I tested only 64-bit. I will test 32-bit about 10 hours later.
And thanks for uploading ming64 build box!
I will try to check it after that.

**Comment by sanak** [comment]
Sorry for late response.
I confirmed that my pgrouting4w native msvc 32-bit build was no problem, too.
I will start to check ming64 build box.

**Comment by sanak** [comment]
I understood this issue's difficulty. (if "DEBUG" equals 1, then the error didn't occur.)
I think that this fix will be later of this month, even if I will be able to have enough time...

**Comment by robe2** [comment]
@sanak 

No problem.  I have a patch in 2.1.0 for now that just causes notice to be enabled where the error occurs.  So it works in 64-bit now as long as people have log notice or show client message set to show at least NOTICE.  We'll just have to make a note of that in docs until the issue is properly fixed.

**Comment by sanak** [comment]
@robe2
Okay, thanks for confirmation and details.

**Comment by sanak** [comment]
Today, I noticed that my pgrouting4w repository `master_msvc` branch's msvc x64 build failed
and after comparing x86 and x64 build difference, I noticed that this issue cause was GraphDefinition class member variable initialization.
(The following image's left side is VS2013 x86 project and right side is x64 project, and bottom center panel initialized values are apparently different.)
![compare_msvc_x86_and_x64](https://cloud.githubusercontent.com/assets/629923/9460838/a9ab04bc-4b44-11e5-9a82-bec1a0759749.png)
I fixed it and confirmed that mingw64, msvc x86 and x64 and Mac OSX homebrew builds are succeeded,
so could someone check my pull request (https://github.com/pgRouting/pgrouting/pull/402) and test by executing `tools/test-runner.pl -ignorenotice` ?
(If possible, I hope that the original TRSP developer (in my memory, GSoC student ?) can check my pull request (https://github.com/pgRouting/pgrouting/pull/402).)
Thanks,

**Comment by sanak** [comment]
@robe2 
I fixed GraphDefinition initialization bug, and it was merged to develop_2_1_0 by @cvvergara.
So, I think that your following temporary fix is not necessary anymore.
https://github.com/pgRouting/pgrouting/commit/3edbc88109588e30df965a60c3556c9a6ecb5ed6

**Comment by sanak** [comment]
@cvvergara, @robe2
I think that the following is necessary.
<pre>
git revert 3edbc88
</pre>

**Comment by cvvergara** [comment]
Why? is it causing a problem?

**Comment by sanak** [comment]
https://github.com/pgRouting/pgrouting/issues/336#issuecomment-129589774
> We'll just have to make a note of that in docs until the issue is properly fixed.

Yes, missing GraphDefinition initialization caused this problem and I fixed it properly.

**Comment by cvvergara** [comment]
ok, its fixed with those initializations, but then the @robe2  temp fix, can be removed?

**Comment by sanak** [comment]
Yes, I think that it can be removed.
About master branch, I confirmed that it fixed the issue on my local ming64gcc48 environment by #403 which doesn't include temporary fix.

**Comment by sanak** [comment]
Sorry, again.
But, isn't the following necessary in Release 2.1.0 ?
<pre>
git revert 3edbc88
</pre>
https://github.com/pgRouting/pgrouting/commit/3edbc88109588e30df965a60c3556c9a6ecb5ed6

**Comment by robe2** [comment]
@sanak I thought you just said your fixes to trsp made my hack not necessary anymore? 

But let me check -- I broke something on winnie, so she's having trouble testing.

**Comment by robe2** [comment]
@sanak   I talked with Vicky.  She doesn't want to make any more changes to 2.1 and doesn't want to risk taking the hack out.  So she wants to do it on 2.2.  I'm fine with that.  We can do it immediately after we release 2.1.

**Comment by sanak** [comment]
@robe2 
> I thought you just said your fixes to trsp made my hack not necessary anymore?

Yes.

@robe2, @cvvergara 
> I talked with Vicky. She doesn't want to make any more changes to 2.1 and doesn't want to risk taking the hack out. So she wants to do it on 2.2. I'm fine with that. We can do it immediately after we release 2.1.

Okay, no problem.
Thanks for confirmation!

**Comment by cvvergara** [comment]
TRSP might get a total rewrite on V2.3, so I repoened so we dont forget that this issue is not forgotten for testing

**Comment by dkastl** [comment]
What about opening a new issue instead with a reference to this one?
Might be easier to follow.

**Comment by robe2** [comment]
Yah I think that would be easier and people running 2.1 won't think the issue is not fixed.

**Comment by cvvergara** [comment]
Looks like it was fixed on V2.1, but the milestone for 2.1 is closed, So I will tag as fixed in develop and close it.
The label name will help find it by label.
